### ISSUE TITLE
[Feature][Request] Support HeteroGraph for RelGraphConv module

### ISSUE THREAD

**Issue by chang-l** [dmlc/dgl#4646]
## 🚀 Feature

When it comes to using RelGraphConv module, it seems we always converted to a homogeneous graph for all current examples; while for hetero-graph RGCN, we used another more general module (HeteroGraphConv). 

For small problems, converting to homo. graph wouldn't add too much overhead of using RelGraphConv; however, for building larger Hetero-RGCN modules with many relation types, can we also support RelGraphConv-like implementation, meaning, issuing all relations in a batched fashion or a single `segment_MM` operation, instead of iterating over every relation type (HeteroGraphConv)? 

## Motivation

With the sacrifice of generality of HeteroGraphConv, this, however, can reduce user's effort of implementing RGCN with hetero-graph, and improve GPU utilization when every single relation is not so intensive.

## Alternatives
For hetero-RGCN, current alternatives are: (1) `to_homogeneous` + `RelGraphConv`, which performs well but only for homogeneous graph; and (2) `HeteroGraphConv`, which is very flexible and customizable, but may be less convenient and efficient.

## Pitch

A RelGraphConv module that can support hetero-graph with single conv but many relation types, for better performance.

**Comment by jermainewang** [comment]
Hi @chang-l , this thread was picked up by our bot as there have not been many activities. I could follow up my thought. My view of the current alternatives (1) and (2) is that they are different kinds of GNN layers. `RelGraphConv` assumes that the embeddings of different node types have the same size and data type. And because of this, `RelGraphConv` can implement advanced regularization methods such as basis and bdd regularization. On the other hand, `HeteroGraphConv` does not have this restriction -- users can pass in completely different embeddings for different node types. Although `HeteroGraphConv` + `GraphConv` happens to be similar to `RelGraphConv`, they are inherently designed for different user scenarios.

In terms of the future roadmap, I think we are likely to continue to keep and optimize both options. For (1), we now know that `segment_mm` or `gather_mm` kernels can speedup the computation side, then the bottleneck may shift to data loading stage (e.g., how to efficiently prepare mini-batch graph for `RelGraphConv`). For (2), how to automatically batch a user customized `HeteroGraphConv` will be an interesting topic. Perhaps we can look at some compilation techniques for autobatching.

**Comment by chang-l** [comment]
@jermainewang 
I agree that `RelGraphConv` and `HeteroGraphConv` are suitable for different scenarios, and they are both necessary to ensure the flexibility of DGL. Regarding the future roadmap, let me rephrase this request (as I also understand it better now) for discussion: 

(1) For `RelGraphConv`, this request is not about performance.  I am wondering if we can also support some hertero-graph `g` (same feature size for all types) as one of the input of `RelGraphConv` module? Because, naturally, `RelGraphConv` deals with heterogeneous graph, and most likely user would have heterogeneous graph in the first place. Asking them to convert to homogeneous graph whenever using `RelGraphConv` might be inconvenient/error-pruned. Is it possible to hide `to_homogeneous()` or trim some preprocessing effort (like [here](https://github.com/dmlc/dgl/blob/ed8e9c4440c204d19a138bcc552ee476c3a7cbc1/examples/pytorch/rgcn/entity_sample.py#L98-L107))? Would it introduce significant overheads?

(2) For `HeteroGraphConv`, yes, there is more room to improve the performance. Hopefully, those different sized small MMs can be automatically fused by some (GNN)compilers in future :)

**Comment by jermainewang** [comment]
> Is it possible to hide to_homogeneous() or trim some preprocessing effort (like [here](https://github.com/dmlc/dgl/blob/ed8e9c4440c204d19a138bcc552ee476c3a7cbc1/examples/pytorch/rgcn/entity_sample.py#L98-L107))?

You've pointed out a valid issue in our current implementation. If the preprocess step is only one `to_homogeneous()` call, I believe the burden is acceptable but it apparently is not the case. My feeling is that there are multiple issues mingled together here. Let's break them down:

https://github.com/dmlc/dgl/blob/7c059e86c736f19b63d14df789e546ce479dfdfa/examples/pytorch/rgcn/entity_sample.py#L97-L101

L97-L101 calls `to_homogeneous()`, gets the target category ID and the node IDs belonging to the category. Among them only L99 is caused by `RelGraphConv` while others are needed by mini-batch training.

https://github.com/dmlc/dgl/blob/7c059e86c736f19b63d14df789e546ce479dfdfa/examples/pytorch/rgcn/entity_sample.py#L102-L104

L102-L104 seems to be nuisance caused by our DataLoader. Is it possible to fix them?

https://github.com/dmlc/dgl/blob/7c059e86c736f19b63d14df789e546ce479dfdfa/examples/pytorch/rgcn/entity_sample.py#L105-L107

My understanding of L105-L107 is because although the graph has been converted to homogeneous, the node labels are stored as type-wise tensors. Its length is the number of the nodes of the target category instead of the number of total nodes. Therefore, the code creates a mapping for fetching the correct labels. I wonder what if our `to_homogeneous()` API can pad the label tensor to the right size automatically? Will this simplify the logic?

cc @mufeili
